2019-11-05-parse-don-t-validate (#55)
rust-lang-ru/translations: ru/articles/2019-11-05-parse-don-t-validate.md

```
Title: Парсите, а не валидируйте
Date: 2019-11-05T17:09:58

> Tags: Функциональное программирование, хаскель, типы

Reviewer: ```suggestion
Tags: Функциональное программирование, хаскель, типы ffff
```

    }
}

fn main() -> Result<(), &'static str> {
    let config_dirs =get_configuration_directories()?;
    initialize_cache(head(config_dirs));
    Ok(())
}
```

Заметьте, что лишняя проверка в  `main` полностью исчезла! Вместо этого, мы производим проверку ровно один раз, в функции `get_configuration_directories`. Она конструирует `NonEmptyVec` из `Vec` с помощью функции  `non_empty`, имеющую следующую сигнатуру:

```rust
fn non_empty<T>(list: Vec<T>) -> Option<NonEmptyVec<T>>
```

Обратите внимание, что `Option` никуда не делся, однако в этот раз мы обрабатываем случай `None` в самом начале программы: в том месте где мы и раньше делали валидацию. Когда проверка пройдена, мы получаем значение типа  `NonEmptyVec`, которое сохраняет (на уровне типов!) тот факт, что список на самом деле не пуст. Другими словами, значение типа `NonEmptyVec` эквивалентно значению типа `Vec<T>` плюс *доказательству*, что список не пуст.

Усилив тип аргумента функции `head` вместо того чтобы ослабить тип результата, мы полностью избавились от всех проблем из предыдущего раздела:

- В коде нет лишних проверок, поэтому нет и никакого оверхеда по производительности

> - Более того, если функция `get_configuration_directories` перестанет проверять список на пустоту, то её результирующий тип тоже поменяется. Следовательно, функция `main` не сможет тайпчекнуться, сообщая о проблеме до того как мы вообще запустили программу!

Reviewer 2: ```suggestion
 - Ликвидация проверки на пустоту в функции `get_configuration_directories()` повлечёт изменения в её результирующем типе, следовательно, функция `main()` не сможет тайп-чекнуться, сообщая о проблеме до того, как мы вообще запустили программу!
```

> Более того, мы можем тривиально восстановить старое поведение функции  `head` с помощью новой версии, композируя её с `non_empty`:

Reviewer 2: ```suggestion
Более того, мы можем тривиально восстановить старое поведение функции `head()` с помощью новой версии, композируя её с `non_empty()`:
```

```rust
fn old_head<T>(xs: Vec<T>) -> Option<T> {
    non_empty(xs).map(head)
}
```

> Обратите внимание, что обратное *неверно*: не существует способа получить новую версию функции `head` из старой. Таким образом, второй подход превосходит первый по всем параметрам.

Reviewer 2: ```suggestion
Обратите внимание, что обратное *неверно*: не существует способа получить новую версию функции `head()` из старой. Таким образом, второй подход превосходит первого по всем параметрам.
```

## Сила парсинга

> Возможно, вы задаётесь вопросом, какое отношение это всё имеет к заголовку статьи. В конце концов, мы просто изучили два разных способа проверить список на пустоту — и на первый взгляд, тут нет никакого парсинга. Эта интерпретация не неверна, однако я предлагаю посмотреть на это с другой стороны: с моей точки зрения, вся разница между валидацией и парсингом полностью состоит в том, как сохраняется информация об этом процессе. Давайте сравним две такие функции:

Reviewer 2: ```suggestion
Но какое это всё имеет отношение к заголовку статьи? В конце концов, мы просто изучили два разных способа проверить список на пустоту — и, на первый взгляд, тут нет никакого парсинга. Эта интерпретация неверна, однако я предлагаю посмотреть на это с другой стороны: с моей точки зрения, вся разница между валидацией и парсингом полностью состоит в том, как сохраняется информация об этом процессе. Давайте сравним две такие функции:
```

```rust
fn validate_non_empty<T>(xs: Vec<T>) -> Result<(), UserError> {
    if !xs.is_empty() {
        Ok(())
    } else {
        Err(UserError::new("list cannot be empty"))
    }
}

fn parse_non_empty<T>(mut xs: Vec<T>) -> Result<NonEmptyVec<T>, UserError> {
    if !xs.is_empty() {
        let head = xs.remove(0);
        Ok(NonEmptyVec(head, xs))
    } else {
        Err(UserError::new("list cannot be empty"))
    }
}
```

> Эти две функции практически идентичны: они проверяют переданный список на пустоту, и если он пустой, то они возвращают сообщение об ошибке. Вся разница заключается в возвращаемом значении: `validate_non_empty`всегда возвращает `()`, тип, который не содержит никакой информации, а `parse_non_empty` возвращает `NonEmptyVec<T>`, уточнение входного типа которое сохраняет полученное знание в системе типов. Обе функции проверяют одно и то же, но `parse_non_empty` даёт вызывающему коду доступ к полученной информации, а `validate_non_empty` просто выкидывает её.

Reviewer 2: ```suggestion
Эти две функции практически идентичны: они проверяют переданный список на пустоту, и, если он пустой, они возвращают сообщение об ошибке. Вся разница заключается в возвращаемом значении: `validate_non_empty()` всегда возвращает `()`, тип, который не содержит никакой информации, а `parse_non_empty()` - `NonEmptyVec<T>`, уточнение входного типа, сохраняющее полученное знание в системе типов. Обе функции проверяют одно и то же, но `parse_non_empty()` даёт вызывающему коду доступ к полученной информации, а `validate_non_empty()` просто выкидывает её.
```

> Эти две функции элегантно иллюстрируют два различных взгляда на роль системы типов: `validate_non_empty` просто подчиняется тайпчекеру, но только `parse_non_empty` полностью использует те преимущества, которые он даёт. Если вы видите, почему функция `parse_non_empty` предпочтительнее, то вы должны уже понимать, что означает мантра "парсите, а не валидируйте". Однако, возможно вы скептически относитесь к имени `parse_non_empty`. Действительно ли она что-то *парсит*, или она просто валидирует вход и возвращает результат? И, хотя точное определение того, что означает парсинг или валидация, является предметом для обсуждения, я считаю что `parse_non_empty` это полноценный парсер, пусть и очень простой.

Reviewer 2: ```suggestion
Эти две функции элегантно иллюстрируют два различных взгляда на роль системы типов: `validate_non_empty()` просто подчиняется тайп-чекеру, но только `parse_non_empty()` полностью использует те преимущества, которые он даёт. Если вы видите, почему функция `parse_non_empty()` предпочтительнее, то вы должны уже понимать, что означает мантра "парсите, а не валидируйте". Однако, возможно вы скептически относитесь к имени `parse_non_empty`. Действительно ли она что-то *парсит*, или она просто валидирует вход и возвращает результат? И, хотя точное определение того, что означает парсинг или валидация, является предметом для обсуждения, я считаю что `parse_non_empty` это полноценный парсер, пусть и очень простой.
```

2. **Выносите бремя проверок как можно выше, но не дальше.** Переведите ваши данные в наиболее точное представление так рано, как это возможно. В идеале, это должно происходить на границе вашей системы, до того как *любые* данные начали обрабатываться.

    Если более точное представление данных нужно только в одной конкретной ветке кода, то парсите данные как только эта ветвь была взята. Используйте тип-суммы с умом, чтобы ваши типы данных отражали возможные результаты выполнения кода.

Другими словами, пишите функции с типами данными, которые вы
*хотели бы* иметь, а не те, которые у вас есть. Процесс проектирования тогда становится упражнением в заполнении этого различия, над которым часто нужно работать с обоих концов, пока вы не сойдётесь на каком-то компромиссном решении. Не бойтесь итеративно изменять части программы , ведь в итоге в процессе рефакторинга вы можете даже узнать что-то новое!

Вот несколько дополнительных советов, расположенных в произвольном порядке:

> - **Позвольте вашим типам данных информировать ваш код, не позволяйте коду контролировать типы данных.** Боритесь с искушением запихнуть `bool` в структуру потому что он нужен в функции, которую вы пишете. Не бойтесь рефакторить код, чтобы использовать более правильное представление данных - система типов удостоверится, что вы учли все места, которые требуют изменения, и это спасёт вас от головной боли в дальнейшем
> 
> - **Относитесь к функциям без возвращаемого результата или с типом `Result<(), Error>` с большим подозрением.** Иногда они совершенно необходимы, потому что они могут производить императивный эффект без какого-либо разумного результата, но если единственная цель такой функции это вызвать ошибку, скорее всего есть более хороший путь.
> 
> - **Не бойтесь парсить данные в несколько этапов.** Избегание парсинга наугад всего лишь означает, что вы не должны работать с выходными данными до того, как полностью их распарсите, а не то, что вы не можете использовать часть входных данных, чтобы решить, как парсить остальное. Множество полезных парсеров контекстно-зависимы.
> 
> - **Избегайте денормализованного представления данных, *особенно*, если они изменяемые** Дублирование одних и тех же данных в множестве мест позволяет элементарно привести систему в некорректное состояние: данные в разных местах могут рассинхронизироваться. Стремитесь к единственному источнику истины
> 
>     - **Держите денормализованные данные за границей абстракций.** Если денормализованное состояние абсолютно необходимо, используйте инкапсуляцию, чтобы существовал только небольшой доверенный модуль, единственная роль которого заключается в поддержании данных в синхронизированном состоянии.
> 
> - **Используйте абстрактные типы данных, чтобы ваши валидаторы "выглядели как" парсеры.** Иногда, сделать данные действительно непредставимыми совершенно непрактично, учитывая возможности которые даёт ваш язык программирования, например, проверка, что число попадает в некоторый диапазон. В этом случае, используйте `декоратор (newtype)` с умным конструктором, чтобы сделать псевдо-парсер из валидатора.
> 
> Как обычно, используйте здравый смысл. Иногда просто не стоит ломать весь код и переписывать всё приложение чтобы избежать единственного возможного `error "impossible"` где-то в коде - просто не забывайте относиться к таким местам как к радиоактивной субстанции, которой они и являются, и обходитесь с ними со всей необходимой аккуратностью. В самом крайнем случае, оставьте хотя бы комментарий, чтобы задокументировать инварианты для тех, кто в будущем будет модифицировать этот код.

Reviewer 2: ```suggestion
- **Позвольте типам данных информировать ваш код, не позволяйте коду контролировать типы данных.** Боритесь с искушением запихнуть `bool` в структуру, потому что он нужен в функции, которую вы пишете. Не бойтесь рефакторить код, чтобы использовать более правильное представление данных - система типов удостоверится, что вы учли все места, которые требуют изменения, что спасёт вас от головной боли в дальнейшем.

- **Относитесь к функциям без возвращаемого результата (или с типом `Result<(), Error>`) с большИм подозрением.** Иногда они совершенно необходимы, потому что они могут производить императивный эффект без какого-либо разумного результата, но если единственная цель функции - вызвать ошибку, скорее всего, есть путь получше.

- **Не бойтесь парсить данные в несколько этапов.** Избегание парсинга наугад всего лишь означает, что вы не должны работать с выходными данными до того, как полностью их распарсите, а не то, что вы не можете использовать часть входных данных, чтобы решить, как парсить остальное. Множество полезных парсеров контекстно-зависимы.

- **Избегайте денормализованного представления данных, *особенно* если они изменяемые**. Дублирование одних и тех же данных в множестве мест позволяет элементарно привести систему в неправильное состояние: данные в разных местах могут рассинхронизироваться. Стремитесь к единственному источнику истины.

 - **Держите денормализованные данные за границей абстракций.** Если денормализованное состояние абсолютно необходимо, используйте инкапсуляцию, чтобы существовал только небольшой доверенный модуль, единственная роль которого заключается в поддержании данных в синхронизированном состоянии.

- **Используйте абстрактные типы данных, чтобы ваши валидаторы "выглядели как" парсеры.** Иногда сделать данные действительно непредставимыми совершенно непрактично, учитывая возможности, которые даёт ваш язык программирования, например, проверка, что число попадает в некоторый диапазон. В этом случае используйте `декоратор (newtype)` с умным конструктором, чтобы сделать псевдо-парсер из валидатора.

Как обычно, используйте здравый смысл. Иногда просто не стоит ломать весь код и переписывать всё приложение, чтобы избежать единственного возможного `error "impossible"` где-то в коде - просто не забывайте относиться к таким местам как к радиоактивной субстанции, которой они и являются, и обходитесь с ними со всей необходимой аккуратностью. В самом крайнем случае, оставьте хотя бы комментарий, чтобы задокументировать инварианты для тех, кто в будущем будет модифицировать этот код.
```

# Резюме

> Вот и всё, в общем-то. Надеюсь, эта статья об использовании преимуществ системы типов не требует докторской степени для понимания, и они даже не требуют последних и крутейших фишек из более мощных систем типов - хотя иногда они сильно могут помочь! Иногда, наибольшим препятствием для использования системы типов на полную катушку является простое незнание, что это возможно.
> 
> Ни одна из идей этой статьи не нова. На самом деле, основная идея - "пишите тотальные функции" - концептуально очень проста. Несмотря на это, удивительно сложно придумать практичные, понятные объяснения того, как я пишу код в таком стиле. Можно легко потратить кучу времени, разговаривая об абстрактных концепциях - многие из которых очень ценны! - не сообщив ничего полезного о *процессе*. Я надеюсь, что эта статья - это небольшой шажок в этом направлении.
> 
> К сожалению, я знаю не так много ресурсов на эту тему, но я знаю одно: я никогда не стесняюсь порекомендовать фантастическую статью Мэта Парсона  [Type Safety Back and Forth](https://www.parsonsmatt.org/2017/10/11/type_safety_back_and_forth.html). Если вам нужна ещё одна доступная статью на тему этих идей, то я крайне рекомендую прочитать её. Для намного более серьёзного погружения в эти идеи я могу порекомендовать работу Мэта Нунана [Ghosts of Departed Proofs](https://kataskeue.com/gdp.pdf), который описывает некоторые более продвинутые техники по выражению более сложных инвариантов, нежели я привёл в статье.
> 
> В заключение, я хотел бы заметить, что рефакторинг в стиле, который описан в статье, не всегда прост. Примеры которые я использовал - просты, но реальная жизнь зачастую куда менее прямолинейна. Даже люди, имеющие большой опыт в  type-driven проектировании, могут испытывать затруднения в выражении некоторых инвариантов в системе типов, так что не падайте духом, если у вас не получается решить проблему так, как вам хочется. Считайте эти принципы идеалом, к которому надо стремиться, а не строгому обязательному требованию. Всё что требуется - хотя бы попытаться.

Reviewer 2: ```suggestion
Вот и всё, в общем-то. Надеюсь, эта статья об использовании преимуществ системы типов не требует докторской степени для понимания, и даже не требует последних и крутейших фишек из более мощных систем типов - хотя иногда они сильно могут помочь! Иногда наибольшим препятствием для использования системы типов на полную катушку является простое незнание, что это возможно.

Ни одна из идей этой статьи не нова. На самом деле, основная идея - "пишите тотальные функции" - концептуально очень проста. Несмотря на это, удивительно сложно придумать практичные, понятные объяснения того, как я пишу код в таком стиле. Можно легко потратить кучу времени, разговаривая об абстрактных концепциях - многие из которых очень ценны! - не сообщив ничего полезного о *процессе*. Я надеюсь, что эта статья - это небольшой шажок в этом направлении.

К сожалению, я знаю не так много ресурсов на эту тему, но я знаю одно: я никогда не стесняюсь порекомендовать фантастическую статью Мэта Парсона  [Type Safety Back and Forth](https://www.parsonsmatt.org/2017/10/11/type_safety_back_and_forth.html). Если вам нужна ещё одна доступная статью на тему этих идей, то я крайне рекомендую прочитать её. Для намного более серьёзного погружения в эти идеи я могу порекомендовать работу Мэта Нунана [Ghosts of Departed Proofs](https://kataskeue.com/gdp.pdf), в которой изложены некоторые более продвинутые техники по выражению более сложных инвариантов, нежели чем те, что я привела в статье.

В заключение, я хотел бы заметить, что рефакторинг в стиле, который описан в статье, не всегда прост. Примеры которые я использовал - просты, но реальная жизнь зачастую куда менее прямолинейна. Даже люди, имеющие большой опыт в TypeDD, могут испытывать затруднения в выражении некоторых инвариантов в системе типов, так что не падайте духом, если у вас не получается решить проблему так, как вам хочется. Считайте эти принципы идеалом, к которому надо стремиться, а не строгому обязательному требованию. Всё, что требуется - хотя бы попытаться.
```

Однако, около месяца назад, [я рассуждал в Твиттере] о различии парсинга JSON в статических и динамических языках, и, наконец, я нашёл то, что искал. И теперь у меня есть единственный, бойкий (?) слоган, который является квинтэссенцией того, чем для меня является type-driven design, и что ещё лучше, оно состоит всего лишь из трёх слов:

<div style="text-align: center; font-size: larger"><strong>Парсите, а не валидируйте.</strong></div>

<!-- more -->

# Суть type-driven design

Ладно, признаю: если только вы не знаете уже что такое type-driven design то мой хитроумный слоган скорее всего ничего для вас не значит. К счастью, именно для этого и написана вся остальная статья. Я собираюсь объяснить что я имею в виду во всех подробностях - но сначала давайте немного пофантазируем.

## Ландшафт возможностей

> Одна из самых замечательных вещей в статической типизации заключается в том, что они позволяют, а иногда и с лёгкостью, ответить на вопросы вроде "а возможно ли вообще написать такую функцию?". Для контрастного примера, давайте посмотрим на такую сигнатуру:

Reviewer 3: говорится о одной вещи а тут `что они позволяют...`

```haskell
enum Void {}

fn foo(x: i32) -> Void
```

Возможно ли реализовать такую функцию `foo`? Очевидно, ответ на этот вопрос - *нет*, так как `Void` - это тип с нулём возможных значений, поэтому невозможно написать *ни одной* функции которая бы возвращала значение типа `Void`. Этот пример довольно скучный, но вопрос становится куда интереснее если мы посмотрим на более реалистичный пример:

```haskell
fn head<T>(xs: Vec<T>) -> T
```

> Эта функция возвращает первый элемент списка. Возможно ли её реализовать? Это совершенно не выглядит как что-то сложное, но если мы попробуем её реализовать, то компилятор не будет доволен:

Reviewer 3: `компиляор будет недоволен`

> ```haskell

Reviewer 3: тут надо Rust

### Управление ожиданиями

Как мы убедились, функция `head` является частичной, потому что невозможно вернуть элемент из пустого списка: мы делаем обещание, которое невозможно соблюсти. К счастью, существует простое решение этой дилеммы: мы можем ослабить наше обещание. Так как мы не можем гарантировать, что в списке будет элемент который мы могли бы вернуть, мы можем немного повлиять на ожидания вызывающего кода: мы сделаем всё возможное, чтобы вернуть элемент, однако мы оставляем за собой право ничего не вернуть. В Rust, мы выражаем эту возможность при помощи типа `Option`

```haskell
fn head<T>(xs: Vec<T>) -> Option<T>
```

Это даёт нам немного свободы, которой достаточно чтобы реализовать `head` — позволяя вернуть `None`, когда мы понимаем, что не можем вернуть никакого значения типа `T`:

```haskell
fn head<T>(xs: Vec<T>) -> Option<T> {
    match xs.as_slice() {
        [x, ..] => Some(*x),
        [] => None,
    }
}
```

Проблема решена, так? Ну, на текущий момент - да... Однако это решение имеет неявную цену.

> Возвращение `Maybe`, несомненно, удобно, когда мы *реализуем*  `head`. Однако, значительно менее удобно становится её использовать. Так как `head` всегда может вернуть `None`, на весь вызывающий код накладывается бремя проверок этой возможности, и иногда такой перевод стрелок очень напрягает. Чтобы понять почему, давайте посмотрим на такой код:

Reviewer 3: Option?

```haskell
fn get_configuration_directories() -> Result<Vec<String>, &'static str> {
    let config_dirs_string = std::env::var("CONFIG_DIRS").map_err(|_| "cannot read env")?;
    let list: Vec<_> = config_dirs_string.split(',').map(|x| x.to_string()).collect();
    if list.is_empty() {
        return Err("CONFIG_DIRS cannot be empty");
    }
    Ok(list)
}

fn main() -> Result<(), &'static str> {
    let config_dirs =get_configuration_directories()?;
    match head(config_dirs) {
        Some(cacheDir) => initialize_cache(cacheDir),
        None => panic!("should never happen; already checked configDirs is non-empty")
    }
    Ok(())
}
```

> Когда функция `getConfigurationDirectories` получает список путей из окружения, она сразу проверяет что список непустой. Однако, когда мы используем `head` в `main`, чтобы получить первый элемент списка, результат типа `Option<&str>` требует от нас проверки случая `None`, который, как мы знаем, никогда не произойдёт! Это чрезвычайно плохо по многим причинам:

Reviewer 3: надо как в коде `get_configuration_directories`

1. Во-первых это просто утомительно. Мы уже проверили список на пустоту, почему мы должны замусоривать код дополнительными ненужными проверками?

2. Во-вторых, оно имеет потенциальные последствия с точки зрения производительности. И хотя стоимость лишней проверки совсем незначительна в данном конкретном случае, легко можно придумать более сложный сценарий, где ненужные проверки будут накладываться друг на друга, как например в случае если они производятся в маленьком цикле в большим количеством итераций.

> 3. Наконец, и что хуже всего, в этом коде затаился баг. Что, если `getConfigurationDirectories` изменят так, что она перестанет проверять список на пустоту, не важно, случайно или специально? Программист может и не помнить, что нужно обновить `main`, и внезапно "невозможная" ошибка станет не просто возможной, но очень даже вероятной.

Reviewer 3: `get_configuration_directories`

Однако, в том виде как оно сейчас написано, чтобы найти этот баг, мы должны писать тесты или проводить ручную инспекцию кода.

### Платим вперёд

Понятно, что наша модифицированная версия функции `head` работает не совсем так, как нам хотелось бы. Мы хотели бы, чтобы она каким-то образом была умнее: если мы уже проверили, что список не пуст,  `head` должен безусловно вернуть первый элемент, не принуждая нас обрабатывать случай, про который мы точно знаем, что он невозможен. Как бы нам это сделать?

Давайте посмотрим на изначальную (частичную) сигнатуру `head` ещё раз.

```haskell
fn head<T>(xs: Vec<T>) -> T
```

В предыдущем разделе мы превратили частичную сигнатуру в тотальную, ослабив требования к возвращаемому типу. Однако, раз мы не хотим этого, у нас остаётся лишь одна вещь, которую мы можем поменять: тип аргумент (в нашем случае, `Vec<T>`). Вместо того, чтобы ослабить возвращаемый тип, мы можем *усилить* тип аргумента, убрав саму возможность того что  `head` может быть вызвана на пустом списке.

Чтобы это сделать, нам нужен тип, представляющий непустые списки.

> К счастью, такой тип `NonEmptyVec``` несложно написать. У него будет такое определение:

Reviewer 3: разметка md попыла

```haskell
struct NonEmptyVec<T>(T, Vec<T>);
```

> Заметьте, что `NonEmptyVec` всего лишь пара из значения типа `T` и обычного (возможно, пустого) `Vec<T>`. Такая структура удобно моделирует непустой список путём сохранения первого элемента отдельно от хвоста: даже если второй компонент `Vec<T>` представляет собой `[]```, то первый компонент всегда должен присутствовать. Благодаря этому реализация функции `head` становится совершенно тривиальной:

Reviewer 3: разметка md поплыла

После того, как вы это сделаете, место вызова этой функции скорее всего не скомпилируется, потому что в качестве аргумента ей всё ещё
передаётся список пар. Если этот список передаётся в качестве одного из аргументов или если он получен как результат вызова какой-то другой функции, вам нужно продолжать заменять список на `HashMap` по всей цепочке вызовов. В конце концов, вы достигните точки, где это значение создаётся, либо вы найдёте место, где дубликаты должны быть разрешены. В этом месте вы должны вставить модифицированную версию `check_no_duplicate_keys`:

```haskell
fn check_no_duplicate_keys<K: Eq, V>(xs: &[(K,V)]) -> HashMap<K,V> { ... }
```

И теперь проверка *не может* быть удалена, потому что результат функции необходим для того, чтобы программа продолжила работу!

В этом гипотетическом сценарии видны две простые идеи:

> 1. **Используйте структуры данных, которые делают некорректные состояния невозможными.** Моделируйте ваши данные используя наиболее точные структуры данных. Если искоренение некоторой возможности слишком сложно используя текущее представление, попробуйте использовать другое представление, которое позволяет более просто описать свойство, которые вы хотите выразить. Не бойтесь рефакторить

Reviewer 4: > описать свойство, которые вы хотите выразить

свойства, которые

2. **Выносите бремя проверок как можно выше, но не дальше.** Переведите ваши данные в наиболее точное представление так рано, как это возможно. В идеале, это должно происходить на границе вашей системы, до того как *любые* данные начали обрабатываться.

    Если более точное представление данных нужно только в одной конкретной ветке кода, то парсите данные как только эта ветвь была взята. Используйте тип-суммы с умом, чтобы ваши типы данных отражали возможные результаты выполнения кода.

> Другими словами, пишите функции с типами данными которые вы

Reviewer 4: > пишите функции с типами данными которые вы

с типами данных, которые вы

- **Избегайте денормализованного представления данных, *особенно*, если они изменяемые** Дублирование одних и тех же данных в множестве мест позволяет элементарно привести систему в некорректное состояние: данные в разных местах могут рассинхронизироваться. Стремитесь к единственному источнику истины

    - **Держите денормализованные данные за границей абстракций.** Если денормализованное состояние абсолютно необходимо, используйте инкапсуляцию, чтобы существовал только небольшой доверенный модуль, единственная роль которого заключается в поддержании данных в синхронизированном состоянии.

- **Используйте абстрактные типы данных, чтобы ваши валидаторы "выглядели как" парсеры.** Иногда, сделать данные действительно непредставимыми совершенно непрактично, учитывая возможности которые даёт ваш язык программирования, например, проверка, что число попадает в некоторый диапазон. В этом случае, используйте `декоратор (newtype)` с умным конструктором, чтобы сделать псевдо-парсер из валидатора.

> Как обычно, используйте здравый смысл. Иногда просто не стоит ломать весь код и переписывать всё приложение чтобы избежать единственного возможного `error "impossible"` где-то в коде - просто не забывайте относиться к таким местам как к радиоактивной субстанцией, которой они и являются, и обходитесь с ними со всей необходимой аккуратностью. В самом крайнем случае, оставьте хотя бы комментарий, чтобы задокументировать инварианты для тех, кто в будущем будет модифицировать этот код.

Reviewer 4: > относиться к таким местам как к радиоактивной субстанцией

к радиоактивной субстанции

[]

# Резюме

> Вот и всё, в общем-то. Надеюсь, эта статья об использовании преимуществ системы типов не требует докторской степени для понимания, и они даже не требуют последних и крутейших фишек из более мощных систем типов - хотя иногда они сильно могут помочь! Иногда, наибольшим препятствием использовании системы типов на полную катушку является простое незнание, что это возможно.

Reviewer 4: > наибольшим препятствием использовании системы

препятствием в использовании

> Ни одна из идей этой статьи не нова. На самом деле, основная идея - "пишите тотальные функции" - концептуально очень проста. Несмотря на это, удивительно сложно донести практические, практичные подробности о том, как я пишу код в таком стиле. Можно легко потратить кучу времени, разговаривая об абстрактных концепциях - многие из которых очень ценны! - не сообщив ничего полезного о *процессе*. Я надеюсь, что эта статья это небольшой шажок в этом направлении.

Reviewer 3: `практические, практичные` -> `практические`

> Ни одна из идей этой статьи не нова. На самом деле, основная идея - "пишите тотальные функции" - концептуально очень проста. Несмотря на это, удивительно сложно донести практические, практичные подробности о том, как я пишу код в таком стиле. Можно легко потратить кучу времени, разговаривая об абстрактных концепциях - многие из которых очень ценны! - не сообщив ничего полезного о *процессе*. Я надеюсь, что эта статья это небольшой шажок в этом направлении.

Reviewer 4: > Я надеюсь, что эта статья это небольшой шажок

эта статья - это небольшой шажок
